dash-spv/src/sync/headers/manager.rs

```diff
@@ -535,64 +526,50 @@ impl<S: StorageManager + Send + Sync + 'static, N: NetworkManager + Send + Sync
-                        // We have genesis header at index 0
-                        if let Some(genesis_header) =
-                            self.chain_state.read().await.header_at_height(0)
-                        {
-                            tracing::info!(
+        // If this is the first headers2 message, and we need to initialize compression state
+        if !headers2.headers.is_empty() {
+            // Check if we need to initialize the compression state
+            let state = self.headers2_state.get_state(peer_id);
+            if state.prev_header.is_none() {
+                // If we're syncing from genesis (height 0), initialize with genesis header
+                if self.chain_state.read().await.tip_height() == 0 {
+                    // We have genesis header at index 0
+                    if let Some(genesis_header) = self.chain_state.read().await.header_at_height(0)
+                    {
+                        tracing::info!(
                             "Initializing headers2 compression state for peer {} with genesis header",
                             peer_id
                         );
-                            self.headers2_state.init_peer_state(peer_id, *genesis_header);
-                        }
-                    } else if self.chain_state.read().await.tip_height() > 0 {
-                        // Get our current tip to use as the base for compression
-                        if let Some(tip_header) = self.chain_state.read().await.get_tip_header() {
-                            tracing::info!(
+                        self.headers2_state.init_peer_state(peer_id, *genesis_header);
+                    }
+                } else if self.chain_state.read().await.tip_height() > 0 {
+                    // Get our current tip to use as the base for compression
+                    if let Some(tip_header) = self.chain_state.read().await.get_tip_header() {
+                        tracing::info!(
                             "Initializing headers2 compression state for peer {} with tip header at height {}",
                             peer_id,
                             self.chain_state.read().await.tip_height()
                         );
-                            self.headers2_state.init_peer_state(peer_id, tip_header);
-                        }
+                        self.headers2_state.init_peer_state(peer_id, tip_header);
                     }
                 }
             }
+        }
 
-            // Decompress headers using the peer's compression state
-            let headers = match self
-                .headers2_state
-                .process_headers(peer_id, headers2.headers.clone())
-            {
-                Ok(headers) => headers,
-                Err(e) => {
-                    tracing::error!(
+        // Decompress headers using the peer's compression state
+        let headers = match self.headers2_state.process_headers(peer_id, headers2.headers.clone()) {
```

Comment:
```suggestion
        let headers = match self.headers2_state.process_headers(peer_id, &headers2.headers) {
```

You should be able to get away without cloning all the headers here if you change `process_headers`.